File: code-examples/php/crud/cursor.php
<?php
require_once __DIR__ . "/vendor/autoload.php";

// replace this with your MongoDB deployment's connection string
$uri = "mongodb+srv://<user>:<password>@<cluster-url>?retryWrites=true&w=majority";

$client = new MongoDB\Client($uri);

// Asynchronous iteration using cursor methods
function asyncIteration() {
    $collection = $client->test->orders;
    $cursor = $collection->find();
    foreach ($cursor as $document) {
        print_r($document);
    }
}
// Manual iteration using cursor methods
function manualIteration() {
    $collection = $client->test->orders;
    $cursor = $collection->find();
    foreach ($cursor as $document) {
        print_r($document);
    }
}
// Fetch all documents using iterator_to_array method
function fetchAll() {
    $collection = $client->test->orders;
    $cursor = $collection->find();
    $allValues = iterator_to_array($cursor);
    echo count($allValues);
}
// Rewind the cursor
function rewindCursor() {
    $collection = $client->test->orders;
    $cursor = $collection->find();
    $firstResult = iterator_to_array($cursor);
    echo "First count: " . count($firstResult);
    $cursor->rewind();
    $secondResult = iterator_to_array($cursor);
    echo "Second count: " . count($secondResult);
}
// Close a cursor
function closeCursor() {
    $collection = $client->test->orders;
    $cursor = $collection->find();
    $cursor->reset();
}

try {
    asyncIteration();
    manualIteration();
    fetchAll();
    rewindCursor();
    closeCursor();
} finally {
    $client = null;
}
?>

File: code-examples/scala/crud/cursor.scala
import org.mongodb.scala._

// Use MongoClient to create a connection
val mongoClient: MongoClient = MongoClient("mongodb+srv://<user>:<password>@<cluster-url>?writeConcern=majority")

// Connect to a database
val database: MongoDatabase = mongoClient.getDatabase("test")

// Retrieve a collection
val orders: MongoCollection[Document] = database.getCollection("orders")

// Asynchronous iteration by using cursor methods
def asyncIteration(): Unit = {
  val cursor: FindObservable[Document] = orders.find()
  cursor.subscribe((doc: Document) => println(doc.toJson()))
}

// Manual iteration using cursor methods
def manualIteration(): Unit = {
  val cursor: FindObservable[Document] = orders.find()
  cursor.foreach(doc => println(doc.toJson()))
}

// Fetch all documents as an array using the toArray method
def fetchAll(): Unit = {
  val cursor: FindObservable[Document] = orders.find()

  cursor.toFuture().foreach { docs =>
    println(s"The collection contains ${docs.length} documents.")
  }
}


// Rewind the cursor (not directly possible in the MongoDB Scala Driver)
def rewind(): Unit = {}

// Close a cursor (nearly not needed in MongoDB Scala driver)
def close(): Unit = {}

def run(): Unit = {
  asyncIteration()
  manualIteration()
  fetchAll()
  mongoClient.close()
}

run()
